llvm/docs/SYCLFrontEndDeviceCompilation.rst

```diff
@@ -0,0 +1,135 @@
+=================================
+
+.. contents::
+   :local:
+
+.. _sycl_frontend_device_compilation:
+SYCL front-end device compiler parses input source, outlines device part of the
+code, applies additional restrictions on the device code (e.g. no exceptions or
+virtual calls), generates LLVM IR bitcode for the device code only and
+"integration header" which provides information like kernel name, parameters
+order and data type for the runtime library. Multiple LLVM IR bitcodes (in case
+of multiple targets) are packaged into a single object by the
+clang-offload-packager.
+
+An example of front-end device compilation command is shown below:
+
+  .. code-block:: console
+
+    $ clang -cc1 -triple spirv64 -fsycl-is-device test.cpp -o test.bc
+
+Front-end device compilation for SYCL offloading can be split into the following
+components - Device code outlining, SYCL kernel function object lowering,
+Generation of device code diagnostics, and Integration header generation. These
+components are explained in the sections below.
+
+*********************
+Device code outlining
+*********************
+  This component is responsible for identifying and outlining "device code" in the
+  single source.
+  Here is a code example of a SYCL program that demonstrates compiler outlining
+  work:
+
+  .. code-block:: c++
+    :linenos:
+    
+      int foo(int x) { return ++x; }
+      int bar(int x) { throw std::exception{"CPU code only!"}; }
+      // ...
+      using namespace sycl;
+      queue Q;
+      buffer<int, 1> a{range<1>{1024}};
+      Q.submit([&](handler& cgh) {
+        auto A = a.get_access<access::mode::write>(cgh);
+        cgh.parallel_for<init_a>(range<1>{1024}, [=](id<1> index) {
+          A[index] = index[0] * 2 + foo(42);
+        });
+      }
+
+  In this example, the compiler needs to compile the lambda expression passed
+  to the `sycl::handler::parallel_for` method, as well as the function `foo`
+  called from the lambda expression for the device.
+
+  The compiler must also ignore the `bar` function when we compile the
+  "device" part of the single source code, as it's unused inside the device
+  portion of the source code (the contents of the lambda expression passed to the
+  `sycl::handler::parallel_for` and any function called from this lambda
+  expression).
+
+  The current approach is to use the SYCL kernel attribute in the runtime to
+  mark code passed to `sycl::handler::parallel_for` as "kernel functions".
+  The runtime library can't mark foo as "device" code - this is a compiler
+  job: to traverse all symbols accessible from kernel functions and add them to
+  the "device part" of the code marking them with the new SYCL device attribute.
+
+********************************************************
+SYCL kernel function object (functor or lambda) lowering
+********************************************************
+  This component creates an OpenCL kernel function interface for SYCL kernels.
```

Comment:
Is OpenCL relevant for all device targets? I'm under the impression that it is not, for example when targeting nvptx and CUDA.